### ISSUE TITLE
kube-certs.service: Failed with result 'exit-code'

### ISSUE THREAD

**Issue by 123batman** [pires/kubernetes-vagrant-coreos-cluster#222]
Hello!

on ubuntu 16.04
vagrant Installed Version: 1.8.7

```
Bringing machine 'master' up with 'virtualbox' provider...
Bringing machine 'node-01' up with 'virtualbox' provider...
==> master: Running triggers before up...
==> master: 2016-12-02 16:13:55 +0100: setting up Kubernetes master...
==> master: Setting Kubernetes version 1.4.6
==> master: Importing base box 'coreos-alpha'...
==> master: Matching MAC address for NAT networking...
==> master: Checking if box 'coreos-alpha' is up to date...
==> master: Setting the name of the VM: kubernetes-vagrant-coreos-cluster_master_1480691648016_3130
==> master: Pruning invalid NFS exports. Administrator privileges will be required...
==> master: Clearing any previously set network interfaces...
==> master: Preparing network interfaces based on configuration...
    master: Adapter 1: nat
    master: Adapter 2: hostonly
==> master: Forwarding ports...
    master: 22 (guest) => 2222 (host) (adapter 1)
==> master: Running 'pre-boot' VM customizations...
==> master: Booting VM...
==> master: Waiting for machine to boot. This may take a few minutes...
    master: SSH address: 127.0.0.1:2222
    master: SSH username: core
    master: SSH auth method: private key
==> master: Machine booted and ready!
==> master: Setting hostname...
==> master: Configuring and enabling network interfaces...
==> master: Exporting NFS shared folders...
==> master: Preparing to edit /etc/exports. Administrator privileges will be required...
● nfs-server.service - NFS server and services
   Loaded: loaded (/lib/systemd/system/nfs-server.service; enabled; vendor preset: enabled)
   Active: active (exited) since Fri 2016-12-02 14:27:57 CET; 1h 46min ago
 Main PID: 1209 (code=exited, status=0/SUCCESS)
   Memory: 0B
      CPU: 0
   CGroup: /system.slice/nfs-server.service

Dec 02 14:27:57 usernamePC systemd[1]: Starting NFS server and services...
Dec 02 14:27:57 usernamePC systemd[1]: Started NFS server and services.
==> master: Mounting NFS shared folders...
==> master: Setting time zone...
==> master: Running provisioner: file...
==> master: Running provisioner: file...
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running provisioner: file...
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running triggers after up...
==> master: Sanitizing stuff...
Identity added: /home/username/.vagrant.d/insecure_private_key (/home/username/.vagrant.d/insecure_private_key)
==> master: Waiting for Kubernetes master to become ready...
==> master: 2016-12-02 16:23:20 +0100: failed to deploy master within timeout count of 50
==> master: Installing kubectl for the Kubernetes version we just bootstrapped...
Error in configuration: 
* unable to read client-cert /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/admin.pem for default-admin due to open /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/admin.pem: no such file or directory
* unable to read client-key /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/admin-key.pem for default-admin due to open /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/admin-key.pem: no such file or directory
* unable to read certificate-authority /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem for default-cluster due to open /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem: no such file or directory
kubectl already installed in version '1.4.6'. Skipping install.
Configuring environment..
cluster "default-cluster" set.
user "default-admin" set.
context "local" set.
switched to context "local".
```

vagrant ssh master:

```
Last login: Fri Dec  2 16:32:53 GMT-1 2016 from 10.0.2.2 on pts/0
CoreOS valpha (1248.0.0)
Update Strategy: No Reboots
Failed Units: 1
  kube-certs.service
core@master ~ $ 
```

core@master ~ $ journalctl | grep kube-certs

```
Dec 02 16:14:58 master coreos-cloudinit[2075]: 2016/12/02 16:14:58 Writing unit "kube-certs.service" to filesystem
Dec 02 16:14:58 master coreos-cloudinit[2075]: 2016/12/02 16:14:58 Writing file to "/etc/systemd/system/kube-certs.service"
Dec 02 16:14:58 master coreos-cloudinit[2075]: 2016/12/02 16:14:58 Wrote file to "/etc/systemd/system/kube-certs.service"
Dec 02 16:14:58 master coreos-cloudinit[2075]: 2016/12/02 16:14:58 Wrote unit "kube-certs.service"
Dec 02 16:15:11 master coreos-cloudinit[2075]: 2016/12/02 16:15:11 Calling unit command "start" on "kube-certs.service"'
Dec 02 16:15:11 master systemd[1]: kube-certs.service: Main process exited, code=exited, status=1/FAILURE
Dec 02 16:15:11 master systemd[1]: kube-certs.service: Unit entered failed state.
Dec 02 16:15:11 master systemd[1]: kube-certs.service: Failed with result 'exit-code'.
Dec 02 16:15:11 master coreos-cloudinit[2075]: 2016/12/02 16:15:11 Result of "start" on "kube-certs.service": failed
Dec 02 16:16:44 master systemd[1]: kube-certs.service: Main process exited, code=exited, status=1/FAILURE
Dec 02 16:16:44 master systemd[1]: kube-certs.service: Unit entered failed state.
Dec 02 16:16:44 master systemd[1]: kube-certs.service: Failed with result 'exit-code'.
```

What else can I do to help finding the issue?

**Comment by 123batman** [comment]
core@master ~ $ sudo find / -name ca.pem
`/etc/kubernetes/ssl/ca.pem`

**Comment by pires** [comment]
Can you try and manually run the script in the master node and share the error?

**Comment by 123batman** [comment]
I think I found part of the problem but now there is a new error.

core@master /tmp $ /usr/sbin/groupadd -r kube-cert
`groupadd: group 'kube-cert' already exists`
core@master /tmp $ /usr/bin/chmod 755 /tmp/make-certs.sh
core@master /tmp $ /tmp/make-certs.sh
```
/etc/kubernetes/ssl/ca-key.pem: Permission denied
139966726739600:error:0200100D:system library:fopen:Permission denied:bss_file.c:402:fopen('/etc/kubernetes/ssl/ca-key.pem','w')
139966726739600:error:20074002:BIO routines:FILE_CTRL:system lib:bss_file.c:404:
```
core@master /tmp $ 
core@master /etc/kubernetes/ssl $ ls -l
```
total 48
-rw-r--r--. 1 root root      1119 Dec  5 16:13 apiserver-csr.pem
-rw-------. 1 root kube-cert 1675 Dec  5 16:13 apiserver-key.pem
-rw-rw----. 1 root kube-cert 1188 Dec  5 16:13 apiserver.pem
-rw-------. 1 root kube-cert 1675 Dec  5 16:13 ca-key.pem
-rw-rw----. 1 root kube-cert 1090 Dec  5 16:13 ca.pem
-rw-r--r--. 1 root root        17 Dec  5 16:13 ca.srl
```

core@master /tmp $ sudo /tmp/make-certs.sh 
```
Generating RSA private key, 2048 bit long modulus
......+++
........+++
e is 65537 (0x10001)
Generating RSA private key, 2048 bit long modulus
.........+++
................................................+++
e is 65537 (0x10001)
Signature ok
subject=/CN=kube-apiserver
Getting CA Private Key
cp: target '/vagrant/artifacts/tls' is not a directory
```

core@master /vagrant $ cd /vagrant/artifacts/
`.gitkeep              hyperkube_v1.4.6.tar  `

The problems seems to be in line 36 in
tls/make-certs-master.sh (I think tls/make-certs-node.sh is fine)

```
# Copy CA stuff to host so worked nodes can use it
cp -p $pem_ca $pem_ca_key /vagrant/artifacts/tls
```
/vagrant/artifacts/tls has not been created yet. 
In line 21
```
# Generate TLS artifacts
mkdir -p "$cert_dir"
```
only
`cert_dir=/etc/kubernetes/ssl`
gets created

**Solution: add /vagrant/artifacts/tls folder to the repo or add mkdir**

I added the tls folder manually. Now I get another error regarding ca.pem. I am going to post it in a new comment.

**Comment by pires** [comment]
If `/vagrant` is not made available to the VMs through NFS, nothing will work.

See https://github.com/pires/kubernetes-vagrant-coreos-cluster#synced-folders.

**Comment by 123batman** [comment]
```
Bringing machine 'master' up with 'virtualbox' provider...
Bringing machine 'node-01' up with 'virtualbox' provider...
==> master: Running triggers before up...
==> master: 2016-12-05 17:06:22 +0100: setting up Kubernetes master...
==> master: Setting Kubernetes version 1.4.6
==> master: Importing base box 'coreos-alpha'...
==> master: Matching MAC address for NAT networking...
==> master: Checking if box 'coreos-alpha' is up to date...
==> master: Setting the name of the VM: kubernetes-vagrant-coreos-cluster_master_1480954004373_40089
==> master: Pruning invalid NFS exports. Administrator privileges will be required...
==> master: Clearing any previously set network interfaces...
==> master: Preparing network interfaces based on configuration...
    master: Adapter 1: nat
    master: Adapter 2: hostonly
==> master: Forwarding ports...
    master: 22 (guest) => 2222 (host) (adapter 1)
==> master: Running 'pre-boot' VM customizations...
==> master: Booting VM...
==> master: Waiting for machine to boot. This may take a few minutes...
    master: SSH address: 127.0.0.1:2222
    master: SSH username: core
    master: SSH auth method: private key
==> master: Machine booted and ready!
==> master: Setting hostname...
==> master: Configuring and enabling network interfaces...
==> master: Exporting NFS shared folders...
==> master: Preparing to edit /etc/exports. Administrator privileges will be required...
● nfs-server.service - NFS server and services
   Loaded: loaded (/lib/systemd/system/nfs-server.service; enabled; vendor preset: enabled)
   Active: active (exited) since Sun 2016-12-04 15:35:17 CET; 1 day 1h ago
  Process: 1131 ExecStart=/usr/sbin/rpc.nfsd $RPCNFSDARGS (code=exited, status=0/SUCCESS)
  Process: 1104 ExecStartPre=/usr/sbin/exportfs -r (code=exited, status=0/SUCCESS)
 Main PID: 1131 (code=exited, status=0/SUCCESS)
   Memory: 0B
      CPU: 0
   CGroup: /system.slice/nfs-server.service

Dec 04 15:35:17 usernamePC systemd[1]: Starting NFS server and services...
Dec 04 15:35:17 usernamePC systemd[1]: Started NFS server and services.
==> master: Mounting NFS shared folders...
==> master: Setting time zone...
==> master: Running provisioner: file...
==> master: Running provisioner: file...
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running provisioner: file...
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running provisioner: shell...
    master: Running: inline script
==> master: Running triggers after up...
==> master: Sanitizing stuff...
Identity added: /home/username/.vagrant.d/insecure_private_key (/home/username/.vagrant.d/insecure_private_key)
==> master: Waiting for Kubernetes master to become ready...
==> master: 2016-12-05 17:09:29 +0100: successfully deployed master
==> master: Installing kubectl for the Kubernetes version we just bootstrapped...
Error in configuration: unable to read certificate-authority /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem for default-cluster due to open /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem: permission denied
kubectl already installed in version '1.4.6'. Skipping install.
Configuring environment..
cluster "default-cluster" set.
user "default-admin" set.
context "local" set.
switched to context "local".
==> master: Configuring Kubernetes DNS...
Error in configuration: unable to read certificate-authority /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem for default-cluster due to open /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem: permission denied
Error in configuration: unable to read certificate-authority /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem for default-cluster due to open /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem: permission denied
```

File permissions of 
/home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem
```
Owner: root
Access read & write
Group: 149
Access: Read & write
Others: None
```
Do you have a suggestion how we / I can take care of this sustainably?

**Comment by 123batman** [comment]
> If /vagrant is not made available to the VMs through NFS, nothing will work.
> 
> See https://github.com/pires/kubernetes-vagrant-coreos-cluster#synced-folders.

Thanks for pointing it out. I am aware of this and double checked the nfs exports. The error is indeed in the missing tls folder as I have explained above. You can see this as I am able to go to /vagrant/artifacts from within the master without problems.

**Comment by pires** [comment]
> Error in configuration: unable to read certificate-authority /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem for default-cluster due to open /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem: permission denied

Can you run `sudo ls -lasR /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/` and paste the result here?

**Comment by 123batman** [comment]
username@usernamePC:~$ sudo ls -lasR /home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/
```
/home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/:
total 556440
     4 drwxrwxr-x 3 username username      4096 Dec  5 17:09 .
     4 drwxrwxr-x 9 username username      4096 Dec  5 18:12 ..
     0 -rw-rw-r-- 1 username username         0 Dec  5 16:44 .gitkeep
556428 -rw------- 1 root     root     569774592 Dec  5 17:08 hyperkube_v1.4.6.tar
     4 drwxrwxr-x 2 username username      4096 Dec  5 17:07 tls

/home/username/vagrant/kubernetes-vagrant-coreos-cluster/artifacts/tls:
total 28
4 drwxrwxr-x 2 username username 4096 Dec  5 17:07 .
4 drwxrwxr-x 3 username username 4096 Dec  5 17:09 ..
4 -rw-r--r-- 1 root     root      891 Dec  5 17:41 admin-csr.pem
4 -rw-r--r-- 1 root     root     1679 Dec  5 17:41 admin-key.pem
4 -rw-r--r-- 1 root     root      977 Dec  5 17:41 admin.pem
4 -rw------- 1 root          149 1679 Dec  5 17:41 ca-key.pem
4 -rw-rw---- 1 root          149 1090 Dec  5 17:41 ca.pem
```
Thank you for your help.

**Comment by pires** [comment]
Why `root`? Those resources should be owned by `username` (assuming `username` is the user that ran `vagrant up`.

**Comment by pires** [comment]
Try again and re-open if it doesn't work.

**Comment by 123batman** [comment]
> Why root?

I had to configure my NFS to preserve the file permissions set inside the VM on the host hence root is the owner, not username. If I don't do this kube-certs.service fails in the line `cp -p $pem_ca $pem_ca_key /vagrant/artifacts/tls` with the error message `cp: failed to preserve ownership for '/vagrant/artifacts/tls/ca.pem': Operation not permitted`. But if I configure NFS to preserve permissions in this way root owns the files

So I have two options here:
1) Leave NFS as is. Problem:
```
core@master /vagrant/tls $ sudo sh make-certs-master.sh
Generating RSA private key, 2048 bit long modulus
........+++
..................................................................+++
e is 65537 (0x10001)
Generating RSA private key, 2048 bit long modulus
.................+++
....................+++
e is 65537 (0x10001)
Signature ok
subject=/CN=kube-apiserver
Getting CA Private Key
cp: failed to preserve ownership for '/vagrant/artifacts/tls/ca.pem': Operation not permitted
cp: failed to preserve ownership for '/vagrant/artifacts/tls/ca-key.pem': Operation not permitted
```

2) Configure NFS to preserve permissions. Problem:
```
Identity added: /home/username/.vagrant.d/insecure_private_key (/home/username/.vagrant.d/insecure_private_key)
==> master: Waiting for Kubernetes master to become ready...
==> master: 2016-12-19 21:55:03 +0100: successfully deployed master
==> master: Installing kubectl for the Kubernetes version we just bootstrapped...
error: unable to read certificate-authority /home/username/vagrant2/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem for default-cluster due to open /home/username/vagrant2/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem: permission denied
kubectl already installed in version '1.5.1'. Skipping install.
Configuring environment..
Cluster "default-cluster" set.
User "default-admin" set.
Context "local" set.
Switched to context "local".
==> master: Configuring Kubernetes DNS...
error: unable to read certificate-authority /home/username/vagrant2/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem for default-cluster due to open /home/username/vagrant2/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem: permission denied
error: unable to read certificate-authority /home/username/vagrant2/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem for default-cluster due to open /home/username/vagrant2/kubernetes-vagrant-coreos-cluster/artifacts/tls/ca.pem: permission denied
```
Either way there are permission problems. I'm wondering how you work around the "failed to preserve ownership"-problem? Or why does it not surface on your system? The same problem comes up with containers which rely on special permissions as well.

**Comment by pires** [comment]
I use MacOS and not Ubuntu. I'm clueless about this situation so I can only hope someone else will run into this and help us figure it out.

**Comment by 123batman** [comment]
Maybe you can post your MacOS NFS default settings so we can compare? What do you think? Here are mine:

cat /etc/idmapd.conf 
```
[General]

Verbosity = 0
Pipefs-Directory = /run/rpc_pipefs
# set your own domain here, if id differs from FQDN minus hostname
# Domain = localdomain

[Mapping]

Nobody-User = nobody
Nobody-Group = nogroup
```

cat /etc/default/nfs-kernel-server
```
# Number of servers to start up
RPCNFSDCOUNT=8

# Runtime priority of server (see nice(1))
RPCNFSDPRIORITY=0

# Options for rpc.mountd.
# If you have a port-based firewall, you might want to set up
# a fixed port here using the --port option. For more information,
# see rpc.mountd(8) or http://wiki.debian.org/SecuringNFS
# To disable NFSv4 on the server, specify '--no-nfs-version 4' here
RPCMOUNTDOPTS="--manage-gids"

# Do you want to start the svcgssd daemon? It is only required for Kerberos
# exports. Valid alternatives are "yes" and "no"; the default is "no".
NEED_SVCGSSD=""

# Options for rpc.svcgssd.
RPCSVCGSSDOPTS=""

# Options for rpc.nfsd.
RPCNFSDOPTS=""
```
cat /etc/default/nfs-common
```
# If you do not set values for the NEED_ options, they will be attempted
# autodetected; this should be sufficient for most people. Valid alternatives
# for the NEED_ options are "yes" and "no".


# Options for rpc.statd.
#   Should rpc.statd listen on a specific port? This is especially useful
#   when you have a port-based firewall. To use a fixed port, set this
#   this variable to a statd argument like: "--port 4000 --outgoing-port 4001".
#   For more information, see rpc.statd(8) or http://wiki.debian.org/SecuringNFS
STATDOPTS=

# Do you want to start the gssd daemon? It is required for Kerberos mounts.
NEED_GSSD=
```
cat /etc/exports
```
# /etc/exports: the access control list for filesystems which may be exported
#		to NFS clients.  See exports(5).
#
# Example for NFSv2 and NFSv3:
# /srv/homes       hostname1(rw,sync,no_subtree_check) hostname2(ro,sync,no_subtree_check)
#
# Example for NFSv4:
# /srv/nfs4        gss/krb5i(rw,sync,fsid=0,crossmnt,no_subtree_check)
# /srv/nfs4/homes  gss/krb5i(rw,sync,no_subtree_check)
#


# VAGRANT-BEGIN: 1000 834418d4-2f11-4b09-ad5b-b274ef72785f
"/home/username/vagrant2/kubernetes-vagrant-coreos-cluster" 172.17.8.101(rw,no_subtree_check,all_squash,anonuid=1000,anongid=1000,fsid=2923918536)
# VAGRANT-END: 1000 834418d4-2f11-4b09-ad5b-b274ef72785f

# VAGRANT-BEGIN: 1000 bae947aa-7a03-4eb5-b8cf-b47a69c46f69
"/home/username/vagrant2/kubernetes-vagrant-coreos-cluster" 172.17.8.102(rw,no_subtree_check,all_squash,anonuid=1000,anongid=1000,fsid=2923918536)
# VAGRANT-END: 1000 bae947aa-7a03-4eb5-b8cf-b47a69c46f69

# VAGRANT-BEGIN: 1000 d24e206f-3c72-462a-9e66-8e087117cffe
"/home/username/vagrant2/kubernetes-vagrant-coreos-cluster" 172.17.8.103(rw,no_subtree_check,all_squash,anonuid=1000,anongid=1000,fsid=2923918536)
# VAGRANT-END: 1000 d24e206f-3c72-462a-9e66-8e087117cffe

```

**Comment by pires** [comment]
I use MacOS defaults, where only `/etc/exports` exists and it's empty.

**Comment by w9n** [comment]
hi, I solved the Error by removing the preserve flag in cp [here](https://github.com/pires/kubernetes-vagrant-coreos-cluster/blob/master/tls/make-certs-master.sh#L36), but then the master is not able to call the nodes.

**Comment by pires** [comment]
Fixed by #233.
